src/orange/orangeinp/detail/DeMorganSimplifier.hh

```diff
@@ -0,0 +1,143 @@
+//---------------------------------------------------------------------------//
+//! \file orange/orangeinp/detail/DeMorganSimplifier.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <utility>
+#include <vector>
+
+#include "orange/orangeinp/CsgTree.hh"
+#include "orange/orangeinp/CsgTypes.hh"
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Implement DeMorgan transformations on a \c CsgTree.
+ *
+ * This class serves as a helper for \c
+ * celeritas::orangeinp::transform_negated_joins implementation. It applies
+ * DeMorgan's law on a \c CsgTree so that all negations of a set operator are
+ * removed and transformed into their equivalent operation.
+ *
+ * The simplification preserves subtrees referred to by \c CsgTree::volumes
+ *
+ * Instances of this class shouldn't outlive the \c CsgTree used to construct
+ * it as we keep a reference to it.
+ *
+ * It is currently single-use: calling operator() twice on the same instance
+ * isn't supported.
+ *
+ * The \c CsgTree being simplified shouldn't contain alias nodes or double
+ * negations.
+ */
+class DeMorganSimplifier
+{
+  public:
+    // Construct a simplifier for the given tree
+    explicit DeMorganSimplifier(CsgTree const&);
+
+    // Perform the simplification
+    CsgTree operator()();
+
+  private:
+    //! CsgTree node 0 is always True{} and can't be the parent of any node
+    //! so reuse that bit to tell that a given node is a volume
+    static constexpr auto is_volume_index{NodeId{0}};
+    //! CsgTree node 1 is always a Negated node parent of node 0, so we can
+    //! reuse that bit to tell if a node has a parent as it's never set for
+    //! node id >= 2
+    static constexpr auto has_parents_index{NodeId{1}};
+
+    //! Helper struct to translate ids from the original tree to ids in the
+    //! simplified tree
+    struct MatchingNodes
+    {
+        //! Set if a node has the exact same node in the simplified tree
+        NodeId unmodified;
+
+        //! Indirections to new simplified join following DeMorgan's law
+        //! Set iif the original node is a negated node.
+        NodeId simplified_to;
+        //! If a join node has been negated, this points to the opposite join
+        //! Set iif the original node is a joined node.
+        NodeId opposite_join;
+
+        //! Set if we need to insert a new negation of that node
+        //! Set iif the original node is a leaf node.
+        NodeId new_negation;
+
+        //! Whether any matching node id is set
+        explicit operator bool() const noexcept
+        {
+            return unmodified || simplified_to || opposite_join || new_negation;
+        }
+
+        // Lookup a node an equal node in the simplified tree
+        NodeId equivalent_node() const;
+    };
+
+    //! Rudimentary 2D square matrix view of a vector<bool>
```

Comment:
This is only used to represent the parent relationship in the original `CsgTree`so the number of nodes never changes. Storage does scale quadratically but at least vector<bool> is very compact. Under a certain number of CsgNode per tree, a sparse matrix could easily use more space since you'd need to store row/col indices though it would still be better to use it. The quadratic memory scaling starts to become significant around 1000 CsgNode (100Mb for that matrix of parents)